design/secure-communication.md

```diff
@@ -0,0 +1,94 @@
+# Secure Communication for Contour and Envoy
+
+Status: Draft
+
+This document outlines a design for securing gRPC communication between Contour and Envoy, including all the xDS endpoints.
+
+## Goals
+
+- Achieve secure communication between Contour and Envoy, which will allow a split deployment.
+- Allow use of the SDS service in Contour with authentication.
+
+## Non Goals
+
+- Certificate management, including creation or rotation. These are left as an exercise for the reader at this time.
+
+## Definitions
+
+### mTLS
+
+In this document, we use **mTLS** as shorthand for 'mutual TLS client certificate authentication', without implying a key lifetime. That is, we're just talking about both sides having TLS certs, not implying that we're using the more modern form which implies dynamic, short-lifetime certificates managed by an automated mechanism.
+
+## Background
+
+Currently, the default deployment of Contour includes a Contour and an Envoy container, that connect via the Pod's loopback interface. This uses the pod-level security boundary to provide transport security (no-one can listen in) and authentication (no-one other than Envoy can connect).
+
+In order to be able to split the deployment of Contour and Envoy, we must be able to provide an acceptable level of both transport security and authentication between the different pods.
+
+Currently, if you do split the deployment, Contour will assume that anything that talks to the xDS endpoints is authorized for all information in them. Notably, this includes any certificates used in any endpoint. Antyhing that can speak SDS can connect to contour and ask for all the certificates it knows about via SDS. This design is intended to mitigate this problem.
+
+It's also important to note that the certificates talked about in this document are **only** the certificates used to secure the gRPC channel between Contour and Envoy, not any certificates used to serve TLS out of Envoy. As such, this design is wholly separate to any certificate functionality on either Ingress or IngressRoute.
+
+## High-Level Design
+
+We will add configuration to Contour, both for `contour serve` and `contour bootstrap`, to have Contour and Envoy use a shared CA and a keypair each. This will allow a basic split deployment with minimal changes to the current code, using simple Secret mounts into file locations in both a Contour Deployment and Envoy deployment or DaemonSet.
+
+The generation of the certificates and their placement into Secrets in an operable and maintainable way is not in scope for this change, only a basic illustration in the example deployment YAMLs and Howto. However, see [Future Work](#future-work) for more info.
+
+## Detailed Design - mTLS
+
+In order for mTLS to work, each end must know:
+- The CA that the TLS relationship will descend from.
+- Its certificate
+- Its key
+
+This applies both to Contour and Envoy.
+
+To accomplish this, we will add the following new command line options:
```

Comment:
Might be nice to also plan about configuring these in the configuration file (https://github.com/heptio/contour/issues/1130)

Reply:
Good catch. I think a note about which flags might move the config file is worth noting. 

IMO not all of these flags will move to the config file, at the moment I'm only expecting that will be for things we pass to `contour serve`. Some of those flags go to `contour bootstrap` and will probably stay there as I'm not super keen in co-mingling `contour bootstrap` configuration cli/flags with `contour serve`s because the former is just an optional helper.

Reply:
I did have a think about this, but figured that we'll just add them to the list of things that go in there when it gets started.